packages/@react-spectrum/tabs/src/Tabs.tsx

```diff
@@ -34,48 +36,130 @@ export function Tabs<T extends object>(props: SpectrumTabsProps<T>) {
     density,
     ...otherProps
   } = props;
+
   let ref = useRef<HTMLDivElement>();
+  let wrapperRef = useRef<HTMLDivElement>();
   let state = useSingleSelectListState<T>({
     ...props,
     onSelectionChange
   });
 
+  let {direction} = useLocale();
   let {styleProps} = useStyleProps(otherProps);
   let {tabListProps, tabPanelProps} = useTabs(props, state, ref);
   let [selectedTab, setSelectedTab] = useState<HTMLElement>();
+  let [collapse, setCollapse] = useState(false);
 
   useEffect(() => {
     let tabs: HTMLElement[] = Array.from(ref.current.querySelectorAll('.' + styles['spectrum-Tabs-item']));
-    setSelectedTab(tabs.find((tab) => tab.dataset.key === state.selectedKey));
-  }, [props.children, state.selectedKey]);
+    let selectedTab = tabs.find((tab) => tab.dataset.key === state.selectedKey);
 
-  return (
-    <div
-      {...styleProps}
-      className={classNames(
-        tabsStyles,
-        'react-spectrum-TabPanel',
-        `react-spectrum-TabPanel--${orientation}`,
-        styleProps.className
-      )}>
+    // default to first tab if selectedTab is deleted from children
+    if (selectedTab != null) {
+      setSelectedTab(selectedTab);
+    } else if (tabs[0]) {
+      state.setSelectedKey(tabs[0].dataset.key);
+      setSelectedTab(tabs[0]);
+    }
+  }, [props.children, state]);
+
+  let onResize = useCallback(() => {
+    let getFurthestPoint = (elem) => {
+      if (direction === 'rtl') {
+        return elem.getBoundingClientRect().left;
+      } else {
+        return elem.getBoundingClientRect().right;
+      }
+    };
+
+    if (wrapperRef.current) {
+      let tabsComponent = wrapperRef.current;
+      let tabs = Array.from(ref.current.children);
+      // The last child is the TabLine so we look at the second to last child for the last Tab
+      let lastTab = tabs[tabs.length - 2];
+
+      let farEdgeTabList = getFurthestPoint(tabsComponent);
+      let farEdgeLastTab = getFurthestPoint(lastTab);
+      let shouldCollapse = direction === 'rtl' ? farEdgeLastTab < farEdgeTabList : farEdgeTabList < farEdgeLastTab;
+
+      if (shouldCollapse) {
+        setCollapse(true);
+      } else {
+        setCollapse(false);
+      }
```

Comment:
Leaving a question in case I forget to bring it up in standup:
The v2 method of determining if tabs should collapse involves keeping the original horizontal Tablist in the DOM even when the Picker is rendered. The Tablist is `aria-hidden` and `visibility: hidden` so it isn't detectable by the user. The above code then checks the position of the farthest tab and compares it to the tabs container's farthest point to determine whether or not to render the Picker.

This is different from the usual approach we have for other components (e.g. ButtonGroup and Breadcrumbs), in which we do a rerender of the component in its original non-collapsed/horizontal position, determine if it is overflowing, and then rerender the component in a collapsed/vertical position if it is overflowing. 

Which approach is preferable? The v2 one is perhaps a bit odd since it has a "extra" component in the DOM when it is collapsed but it doesn't have to do the extra rerenders that the other components need to do.